### DISCUSSION TITLE
Ways to use the same object in multiple async sessions

### DISCUSSION THREAD

**Discussion by tokarenko** [sqlalchemy/sqlalchemy#10897] [Usage Questions]
I have enum-like directory object that I am trying to use in multiple SQLAlchemy async sessions:

```
@mapper_registry.mapped
@dataclass
DirectoryEntry
    id: int
    name: str

    __table__ = Table(
        "directory",
        mapper_registry.metadata,
        Column("id", Integer, primary_key=True),
        Column("name", Text, unique=True),
    )

class Directory:
    ONE = DirectoryEntry(id=1, name="One")
    TWO = DirectoryEntry(id=2, name="Two")
```

The object can't be used directly in multiple sessions: Object '<DirectoryEntry ...>' is already attached to session '2' (this is '3')
The object's copies can be used: one_copy = DirectoryEntry(**asdict(Directory.ONE))

Is it possible to substitute object with its copy on attachment to a session, e.g., via before/after_attach events?

**Comment by CaselIT** [comment]
Hi,

I believe you can use session.merge to achieve this

**Comment by zzzeek** [comment]
yah session.merge() as mentioned will achieve this, using merge() instead of add() will create a copy of that object and associate it with the new session.

**Reply by tokarenko** [reply]
This works, indeed! I have further question as now I am considering to replace all calls to add() with merge() to avoid errors. Will there be additional queries to database (or other performance issues as compared with 'add') on merging of object(s) without PK specified?

**Reply by tokarenko** [reply]
I explored session.merge usage for enum-like objects. As it requires session to do the merge there seem to be two options:

`obj.attr = session.merge(Directory.ONE)`
... or ...
```
obj.attr = Directory.ONE
session.merge(obj)
```
Second option emits SELECT for the whole object to do the merge that is inefficient. I cold not find a way to trigger merge just for changed attributes. Is there any?

First emits only necessary SELECTs but can not be used without a session (and usually objects are created outside of a session). I found two possible solutions for "Object ... is already attached to session error": 
1) expunge() object in after_attach event handler and merge it in before_flush event handler;
2) remove "save-update" cascade to prevent attachment to a session on assignment and then merge() in before_flush event handler;
... all of these options involve SQLALchemy inspection API to find attributes that need to be merged.

Also I tried to mark all enum-like objects as persistent by merging them on application initialization before any assignments. It partially solves the issue but it araises as soon as there are async gather/task groups...

Is there a better way around to replace object with merged copy on assignment/attachment?
